DSharpPlus.Commands/CommandsExtension.cs

```diff
@@ -0,0 +1,233 @@
+namespace DSharpPlus.Commands;
+
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using DSharpPlus.AsyncEvents;
+using DSharpPlus.Commands.EventArgs;
+using DSharpPlus.Commands.Exceptions;
+using DSharpPlus.Commands.Processors;
+using DSharpPlus.Commands.Trees;
+using DSharpPlus.Commands.Trees.Attributes;
+using DSharpPlus.Entities;
+using DSharpPlus.Exceptions;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+/// <summary>
+/// An all in one extension for managing commands.
+/// </summary>
+public sealed class CommandsExtension : BaseExtension
+{
+    /// <inheritdoc cref="CommandsConfiguration.ServiceProvider"/>
+    public IServiceProvider ServiceProvider { get; init; }
+
+    /// <inheritdoc cref="CommandsConfiguration.DebugGuildId"/>
+    public ulong? DebugGuildId { get; init; }
+
+    /// <inheritdoc cref="CommandsConfiguration.UseDefaultCommandErrorHandler"/>
+    public bool UseDefaultCommandErrorHandler { get; init; }
+
+    public CommandExecutor CommandExecutor { get; init; } = new();
+
+    /// <summary>
+    /// The registered commands that the users can execute.
+    /// </summary>
+    public IReadOnlyDictionary<string, Command> Commands { get; private set; } = new Dictionary<string, Command>();
+    private readonly List<CommandBuilder> _commandBuilders = [];
+
+    public IReadOnlyDictionary<Type, ICommandProcessor> Processors { get; private set; } = new Dictionary<Type, ICommandProcessor>();
+    private readonly Dictionary<Type, ICommandProcessor> _processors = [];
+
+    /// <summary>
+    /// Executed everytime a command is finished executing.
+    /// </summary>
+    public event AsyncEventHandler<CommandsExtension, CommandExecutedEventArgs> CommandExecuted { add => this._commandExecuted.Register(value); remove => this._commandExecuted.Unregister(value); }
+    internal readonly AsyncEvent<CommandsExtension, CommandExecutedEventArgs> _commandExecuted = new("COMMANDS_COMMAND_EXECUTED", EverythingWentWrongErrorHandler);
+
+    /// <summary>
+    /// Executed everytime a command has errored.
+    /// </summary>
+    public event AsyncEventHandler<CommandsExtension, CommandErroredEventArgs> CommandErrored
+    {
+        add
+        {
+            if (this.UseDefaultCommandErrorHandler)
```

Comment:
I think we should toggle `this.UseDefaultCommandErrorHandler` when we unregister it